instrumentation/internal/internal-lambda/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/internal/lambda/InnerClassLambdaMetafactoryInstrumentation.java

```diff
@@ -0,0 +1,111 @@
+
+                  @Override
+                  public int mergeReader(int flags) {
+                    return flags;
+                  }
+
+                  @Override
+                  public ClassVisitor wrap(
+                      TypeDescription instrumentedType,
+                      ClassVisitor classVisitor,
+                      Implementation.Context implementationContext,
+                      TypePool typePool,
+                      FieldList<FieldDescription.InDefinedShape> fields,
+                      MethodList<?> methods,
+                      int writerFlags,
+                      int readerFlags) {
+                    return new MetaFactoryClassVisitor(
+                        classVisitor, instrumentedType.getInternalName());
+                  }
+                }));
+  }
+
+  private static class MetaFactoryClassVisitor extends ClassVisitor {
+    private final String slashClassName;
+
+    MetaFactoryClassVisitor(ClassVisitor cv, String slashClassName) {
+      super(Opcodes.ASM7, cv);
+      this.slashClassName = slashClassName;
+    }
+
+    @Override
+    public MethodVisitor visitMethod(
+        int access, String name, String descriptor, String signature, String[] exceptions) {
+      MethodVisitor mv = super.visitMethod(access, name, descriptor, signature, exceptions);
+      // depending on jdk version we instrument either spinInnerClass or generateInnerClass
+      if (("spinInnerClass".equals(name) || "generateInnerClass".equals(name))
+          && "()Ljava/lang/Class;".equals(descriptor)) {
+        mv =
+            new MethodVisitor(api, mv) {
+              @Override
+              public void visitMethodInsn(
+                  int opcode, String owner, String name, String descriptor, boolean isInterface) {
+                super.visitMethodInsn(opcode, owner, name, descriptor, isInterface);
+                // look for a call to ASM ClassWriter.toByteArray() and insert transformation
+                // after it
+                if (opcode == Opcodes.INVOKEVIRTUAL
+                    && "toByteArray".equals(name)
+                    && "()[B".equals(descriptor)) {
+                  mv.visitVarInsn(Opcodes.ALOAD, 0);
+                  mv.visitFieldInsn(
+                      Opcodes.GETFIELD, slashClassName, "lambdaClassName", "Ljava/lang/String;");
+                  mv.visitVarInsn(Opcodes.ALOAD, 0);
+                  // targetClass is used to get the ClassLoader where lambda class will be defined
+                  mv.visitFieldInsn(
+                      Opcodes.GETFIELD, slashClassName, "targetClass", "Ljava/lang/Class;");
+                  mv.visitMethodInsn(
+                      Opcodes.INVOKESTATIC,
+                      Type.getInternalName(LambdaTransformer.class),
+                      "transform",
+                      "([BLjava/lang/String;Ljava/lang/Class;)[B",
+                      false);
+                }
+              }
+            };
+      }
```

Comment:
I ❤️ this.

At the same time I'm a little nervous about it.

I have a lot of confidence in our tests to catch problems, and I'm _not_ worried if this doesn't match/apply in a new Java release, because we'll catch that and fix it similar to when new instrumented library is released that breaks instrumentation.

But I'm a little concerned about old Java versions that we don't test, say old Java 8 versions, or EOL'd Java 9, and this is replacing pretty important instrumentation (though I think it shouldn't break applications if it doesn't match/apply, so that part is good, plus those applications should definitely upgrade their Java versions 😄).

This concern is similar to old instrumented library versions that we don't test, though in those cases we have muzzle range detection to help confidence level.

Maybe a comment linking to the InnerClassLambdaMetafactory git history mentioning that we've manually verified that the conditions needed for this to match/apply are present across all Java 8+ versions(?)

Also it may (or may not) be worth adding Java 8u0 to our test matrix, similar to how we run tests against the oldest instrumented library version that we support.

Let's chat in the SIG meeting later today and see what others think. I can definitely be convinced, would love to not worry anymore about lambdas not getting instrumented (I don't know if we benefit from this anymore in our netty instrumentation given recent work, but it [certainly has caused issues there in the past](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pulls?q=is%3Apr+netty+lambda+is%3Aclosed)).

Reply:
and this is definitely a big concrete win:

> [currently] `ContextStore` in lambda class is backed by shared weak map, so care must be taken that value in map does not reference the key as otherwise there is a memory leak